### PR TITLE
JDK-8310584: GetThreadState reports blocked and runnable for pinned suspended virtual threads

### PR DESCRIPTION
**Pull request by alexmenkov**
The change fixes handling of "suspended" bit in VT state.
The code looks very strange.
java_lang_VirtualThread::RUNNING == 3, so line 803 clears JVMTI_THREAD_STATE_ALIVE(1) and JVMTI_THREAD_STATE_TERMINATED(2)
Per log this code came from loom repo with VT integration.

Testing: tier1-4, updated GetThreadStateMountedTest.java

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8310584](https://bugs.openjdk.org/browse/JDK-8310584): GetThreadState reports blocked and runnable for pinned suspended virtual threads (**Bug** - P3)

### Reviewers
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14878/head:pull/14878` \
`$ git checkout pull/14878`

Update a local copy of the PR: \
`$ git checkout pull/14878` \
`$ git pull https://git.openjdk.org/jdk.git pull/14878/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14878`

View PR using the GUI difftool: \
`$ git pr show -t 14878`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14878.diff">https://git.openjdk.org/jdk/pull/14878.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14878#issuecomment-1634781767)

### CODE REVIEW

**Review by sspitsyn — APPROVED** [review]
This looks good. Thank you for filing bug and fixing it!
I've one question besides this fix.
Thanks,
Serguei

File: src/hotspot/share/prims/jvmtiEnvBase.cpp [hunk 1]
```diff
@@ -800,8 +800,7 @@ JvmtiEnvBase::get_vthread_state(oop thread_oop, JavaThread* java_thread) {
     state = (jint)java_lang_VirtualThread::map_state_to_thread_status(vt_state);
   }
   if (ext_suspended && ((state & JVMTI_THREAD_STATE_ALIVE) != 0)) {
-    state &= ~java_lang_VirtualThread::RUNNING;
-    state |= JVMTI_THREAD_STATE_ALIVE | JVMTI_THREAD_STATE_RUNNABLE | JVMTI_THREAD_STATE_SUSPENDED;
+    state |= JVMTI_THREAD_STATE_SUSPENDED;
   }
```
Comment (sspitsyn) [inline src/hotspot/share/prims/jvmtiEnvBase.cpp:804]: One question unrelated to this bug and your fix.
I wonder if any check and handling is needed for the case:
  `if (ext_suspended && ((state & JVMTI_THREAD_STATE_ALIVE) == 0))`
Not sure this condition is even possible. But do we need to add an assert here?

**Comment by sspitsyn** [comment]
@alexmenkov Do you consider backporting this to 21?

**Review by dholmes-ora — COMMENTED** [review]
The change seems consistent with the definition of `GetThreadState`. But I note that the interrupt bit should also only be set if the target is alive.

File: src/hotspot/share/prims/jvmtiEnvBase.cpp (hunk 1 above)
Comment (alexmenkov) [reply src/hotspot/share/prims/jvmtiEnvBase.cpp:804]: AFAIU it's possible in the case when we have terminated VT and JvmtiVTSuspender is requested to suspend all virtual threads

**Comment by alexmenkov** [comment]
> @alexmenkov Do you consider backporting this to 21?

maybe it makes sense.

**Comment by alexmenkov** [comment]
> The change seems consistent with the definition of `GetThreadState`. But I note that the interrupt bit should also only be set if the target is alive.

we get interrupt bit from Thread object, so the value is consistent with terminated state. suspend bit is a bit different - see my reply to Serguei

File: src/hotspot/share/prims/jvmtiEnvBase.cpp (hunk 1 above)
Comment (dholmes-ora) [reply src/hotspot/share/prims/jvmtiEnvBase.cpp:804]: So there is a window in which is a VT is marked as terminated yet is still visible for actions like this? For regular threads we would always have filtered out thread in the process of exiting. Seeing terminated threads seems potentially problematic but perhaps all the VT code is prepared to handle this.

Comment (alexmenkov) [reply src/hotspot/share/prims/jvmtiEnvBase.cpp:804]: I hope there is no such window in GetThreadState() case, but get_vthread_state method is also called from MultipleStackTracesCollector::fill_frames and there is a comment there:
// Note that either or both of thr and thread_oop
// may be null if the thread is new or has exited.

I keep this check for safety (though fill_frames does not care about suspend bit)

**Comment by dholmes-ora** [comment]
> > The change seems consistent with the definition of `GetThreadState`. But I note that the interrupt bit should also only be set if the target is alive.
> 
> we get interrupt bit from Thread object, so the value is consistent with terminated state. suspend bit is a bit different - see my reply to Serguei

Sorry I don't follow. I don't see anything that prevents the target from terminating after you have read the interrupt bit from the thread object, but before you read the actual state.

**Comment by AlanBateman** [comment]
> Sorry I don't follow. I don't see anything that prevents the target from terminating after you have read the interrupt bit from the thread object, but before you read the actual state.

The virtual thread state and the interrupt status are separate. That's okay for the suspended case, assuming not resumed while JVMTI GetThreadState executes. If not suspended then it looks like it could give an inconsistent view of the state. I don't know why GetThreadState defined a state flag for interrupted.

**Comment by alexmenkov** [comment]
So AFAIU GetThreadState for platform threads (get_thread_state_base) don't have similar issue because suspended/interrupted values are read after reading main thread state value.
For virtual threads suspended/interrupted values are read before. There is a comment in the line 796: `// This call can trigger a safepoint, so thread_oop must not be used after it.` I suppose this is the reason to read them earlier.
Then I think we need to ensure the thread is still alive before applying suspended/interrupted bits

**Review by dholmes-ora — APPROVED** [review]
Seems fine to me, assuming all tests pass.

Thanks.

**Review by sspitsyn — APPROVED** [review]
The update looks good.
Need to make sure there are no new failures though.
Thanks,
Serguei

**Pull request closed without merging**